### PR TITLE
Use a list of expectations to generate a minimal list of WPT variants

### PR DESCRIPTION
**Pull request by kainino0x**
See tools/gen_wpt_cts_html for documentation.

Replaces #92 because I moved the branch into this repo for convenient testing.

### CODE REVIEW

File: src/framework/generate_minimal_query_list.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        const subtree = convertToQuerySplitterTree(v.children, k);
+        queryNode.children.set(k, subtree);
+      }
+    }
+    return queryNode;
+  };
+
+  const log = new Logger();
+  const tree = treeFromFilterResults(log, caselist.values());
+  const queryTree = convertToQuerySplitterTree(tree.children)!;
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply src/framework/generate_minimal_query_list.ts]: Hm, it definitely ended up this way for a reason, but seems to be fixable now.

File: src/tools/gen_wpt_cts_html.ts [hunk 2]
```diff
@@ -1,25 +1,109 @@
...
 import { TestSuiteListingEntry } from '../framework/listing.js';
+import { TestLoader } from '../framework/loader.js';
 import { listing } from '../suites/cts/index.js';
 
-// TODO(kainino0x): Make this file able to take a list of suppressions as input, and a custom output file.
+function usage(rc: number): void {
+  console.error(`\
+Usage:
+  tools/gen_wpt_cts_html TEMPLATE_FILE OUTPUT_FILE [ARGUMENTS_PREFIXES_FILE EXPECTATIONS_FILE EXPECTATIONS_PREFIX SUITE]
+  tools/gen_wpt_cts_html out-wpt/cts.html templates/cts.html
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply src/tools/gen_wpt_cts_html.ts]: hm, I don't think so, we have

```
const outFile = process.argv[2];
const templateFile = process.argv[3];
```

(argv is `['.../node', './tools/gen_wpt_cts_html', ...]`)

File: src/framework/generate_minimal_query_list.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+  const convertToQuerySplitterTree = (
+    children: Map<string, FilterResultTreeNode> | undefined,
+    name?: string
+  ): QuerySplitterTreeNode => {
+    let needsSplit = true;
+
+    if (name !== undefined) {
+      const filter = makeFilter(name);
+      const moreThanOneFile = filter instanceof FilterByGroup;
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply src/framework/generate_minimal_query_list.ts]: idIfSingle alone should probably work. This probably got in here before I had that.

File: src/tools/gen_wpt_cts_html.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (austinEng) [reply src/tools/gen_wpt_cts_html.ts]: huh, then the comment is backwards? 
- The comment is template then output
- the code is output then template

File: src/framework/generate_minimal_query_list.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (kainino0x) [reply src/framework/generate_minimal_query_list.ts]: No wait, that's a single case, test, or file, so possibly both needed. Looking into it

[the remark this answers is not in the document]
Comment (kainino0x) [reply src/framework/generate_minimal_query_list.ts]: Logic was correct, but added a proper function to the abstract class for it.

File: src/tools/gen_wpt_cts_html.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kainino0x) [reply src/tools/gen_wpt_cts_html.ts]: Oh wait, I see, the TEMPLATE_FILE OUTPUT_FILE part is backwards

[the remark this answers is not in the document]
Comment (kainino0x) [reply src/tools/gen_wpt_cts_html.ts]: oops, my github hadn't updated

**Comment by kainino0x** [comment]
PTAAL!

**Comment by kainino0x** [comment]
removed grunt changes, will land separately as they got more complicated

**Pull request merged**